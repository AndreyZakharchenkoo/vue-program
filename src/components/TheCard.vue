<template>
  <div class="card">
    <picture class="card__img">
      <img width="300" height="450" :src="card.poster_path" :alt="card.overview">
    </picture>
    <div class="card__info">
      <h2 class="card__name">{{ card.title }}</h2>
      <p class="card__genre">{{ card.genres[0] }}</p>
      <div class="card__year">{{ card.release_date }}</div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'TheCard',
  props: {
    card: {
      type: Object,
      required: true
    }
  }
}
</script>

<style lang="scss" scoped>
.card {
  img {
    width: 100%;
    height: auto;
    aspect-ratio: 2 / 3;
    object-position: center center;
    object-fit: cover;
  }

  &__info {
    display: grid;
    margin-top: 10px;
    grid-gap: 5px;
    align-items: baseline;
    grid-template-columns: 1fr max-content;
    grid-template-areas:
      'name year'
      'genre genre';
    font-weight: 500;
  }

  &__name {
    grid-area: name;
    margin: 0;
    font-size: 18px;
    opacity: .7;
  }

  &__genre {
    grid-area: genre;
    margin: 0;
    font-size: 14px;
    opacity: .5;
  }

  &__year {
    grid-area: year;
    padding: 5px 10px;
    border: 1px solid #979797;
    opacity: 0.5;
    border-radius: 4px;
  }
}
</style>
