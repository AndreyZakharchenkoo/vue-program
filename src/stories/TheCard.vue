<template>
  <div
    class="card"
    :class="{ extended: isExtended }"
  >
    <router-link
      v-if="!isExtended"
      class="card__link"
      :to="{ name: 'Film', params: { id: card.id }  }"
      @click.native="scrollToTop"
    />
    <picture class="card__img">
      <img
        v-lazyload
        width="327"
        height="590"
        :data-src="card.poster_path"
        :alt="card.title"
        :key="card.poster_path"
      >
    </picture>
    <div class="card__info">
      <h3 class="card__name">{{ card.title }}</h3>
      <p class="card__genre">{{ genre }}</p>
      <div class="card__year">{{ year }}</div>
      <template v-if="isExtended">
        <div v-if="card.vote_average" class="card__rating">{{ card.vote_average }}</div>
        <div v-if="card.runtime" class="card__duration" data-aqa="duration">{{ duration }}</div>
        <div class="card__description">{{ card.overview }}</div>
      </template>
    </div>
  </div>
</template>

<script>

export default {
  name: 'TheCard',
  props: {
    card: {
      type: Object,
      required: true,
      default: () => ({}),
    },
    isExtended: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    genre() {
      const genres = this.card?.genres || [];
      return genres[0];
    },
    year() {
      const year = this.card.release_date || '';
      return year.split('-')[0];
    },
    duration() {
      return `${Math.trunc(this.card.runtime / 60)}h ${this.card.runtime % 60}min`;
    },
  },
  methods: {
    scrollToTop() {
      window.scrollTo({
        top: 0,
        left: 0,
        behavior: 'smooth',
      });
    },
  },
};
</script>

<style scoped>
.card {
  position: relative;
}

.card__img {
  display: block;
  background: url('https://www.lyon-ortho-clinic.com/files/cto_layout/img/placeholder/book.jpg') no-repeat center center;
  background-size: cover;
}

.card__img.loaded img {
  transition: 1s;
  opacity: 1;
}

.card img {
  width: 100%;
  height: auto;
  aspect-ratio: 0.6666666667;
  object-position: center center;
  object-fit: cover;
  opacity: 0;
}

.card__info {
  display: grid;
  margin-top: 20px;
  grid-gap: 5px;
  align-items: baseline;
  grid-template-columns: 1fr max-content;
  grid-template-areas: 'name year' 'genre genre';
  font-weight: 500;
}

.card__name {
  grid-area: name;
  margin: 0;
  font-size: 18px;
  opacity: 0.7;
}

.card__genre {
  grid-area: genre;
  margin: 0;
  font-size: 14px;
  opacity: 0.5;
}

.card__year {
  grid-area: year;
  padding: 2px 10px;
  border: 1px solid #979797;
  border-radius: 4px;
  font-size: 14px;
  opacity: 0.75;
}

.card__rating {
  grid-area: rating;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 70px;
  height: 70px;
  border: 1px solid #fff;
  border-radius: 50%;
  color: #a1e66f;
  font-size: 26px;
}

.card__description {
  grid-area: description;
  font-size: 17px;
  line-height: 1.25;
  opacity: 0.8;
}

.card__duration {
  grid-area: duration;
  color: #f65261;
  font-size: 26px;
  opacity: 0.75;
}

.card__link {
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.card__link:hover ~ .card__info h3 {
  color: #f65261;
}

.card h3 {
  transition: 0.2s;
}

.card.extended {
  display: flex;
  align-items: center;
}

.card.extended .card__img {
  width: 280px;
  margin-right: 60px;
}

.card.extended .card__info {
  flex: 1;
  grid-gap: 20px;
  align-items: baseline;
  grid-template-columns: 100px 150px 1fr 70px;
  grid-template-areas:
    'name name name rating'
    'genre genre genre .'
    'year duration . .'
    'description description description description';
  margin-top: 0;
}

.card.extended .card__genre {
  margin-top: 10px;
}

.card.extended .card__year {
  border: none;
  padding: 0;
  color: #f65261;
  font-size: 26px;
}

.card.extended h3 {
  font-size: 53px;
  line-height: 1.1;
  font-weight: 300;
  opacity: 0.7;
}
</style>
