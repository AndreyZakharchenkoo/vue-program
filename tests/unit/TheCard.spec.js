import { createLocalVue, shallowMount } from '@vue/test-utils'
import TheCard from '@/stories/TheCard.vue'

/*
 * NOTE:
 * Avoiding error with custom directive
 * For correct test executing
 */

import lazyLoadImg from '@/plugins/lazy-loading.js'
const localVue = createLocalVue()
localVue.use(lazyLoadImg)

describe('ThePlaceholder.vue', () => {
  it('Check correct duration transformation', () => {
    const wrapper = shallowMount(TheCard, {
      localVue,
      stubs: ['router-link', 'router-view'],
      propsData: {
        card: {
          id: 353081,
          title: 'Mission: Impossible - Fallout',
          tagline: '',
          vote_average: 0,
          vote_count: 3,
          release_date: '2018-07-26',
          poster_path: 'https://image.tmdb.org/t/p/w500/A2W75CVz9l8gEeWhQiiftlOCkgi.jpg',
          overview: 'The sixth installment in the "Mission: Impossible". When an IMF mission ends badly, the world is faced with dire consequences. As Ethan Hunt takes it upon himself to fulfil his original briefing, the CIA begin to question his loyalty and his motives. The IMF team find themselves in a race against time, hunted by assassins while trying to prevent a global catastrophe.',
          budget: 0,
          revenue: 0,
          genres: [
            'Adventure',
            'Action',
            'Thriller'
          ],
          duration: '1h 24min',
          runtime: null
        },
        isExtended: true
      }
    })

    const duration = wrapper.get('[data-test="duration"]')

    expect(duration.text()).toBe('84 min')
  })
})
