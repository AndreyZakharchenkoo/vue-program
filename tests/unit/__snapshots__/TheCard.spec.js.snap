// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThePlaceholder.vue Renders correctly 1`] = `
<div
  class="card extended"
>
  <!---->
   
  <picture
    class="card__img"
  >
    <img
      alt="Fifty Shades Freed"
      data-src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
      height="590"
      src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
      width="327"
    />
  </picture>
   
  <div
    class="card__info"
  >
    <h3
      class="card__name"
    >
      Fifty Shades Freed
    </h3>
     
    <p
      class="card__genre"
    >
      Drama
    </p>
     
    <div
      class="card__year"
    >
      2018
    </div>
     
    <div
      class="card__rating"
    >
      6.1
    </div>
     
    <div
      class="card__duration"
      data-aqa="duration"
    >
      1h 46min
    </div>
     
    <div
      class="card__description"
    >
      Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.
    </div>
  </div>
</div>
`;
